### PR TITLE
Bump Checkstyle from 8.29 to 8.41 and enable additional checks

### PR DESCRIPTION
**Pull request by basil**
I noticed this warning in the build logs:

```
[INFO] --- maven-checkstyle-plugin:3.1.2:check (validate) @ jenkins-parent ---
[WARNING] Old version of checkstyle detected. Consider updating to >= v8.30
[WARNING] For more information see: https://maven.apache.org/plugins/maven-checkstyle-plugin/examples/upgrading-checkstyle.html
```

Following those instructions, I upgraded [Checkstyle](https://checkstyle.sourceforge.io/) to the latest version (8.41). See the [release notes](https://checkstyle.sourceforge.io/releasenotes.html).

With that having been done, I wanted to enable some checks to ensure that recent cleanup changes do not regress in the future. I enabled a dozen or two rules that we either fully complied with already or mostly complied with already (fixing any existing violations). There are more rules that could be enabled, but I thought this was a good start for now.

### CODE REVIEW

File: pom.xml
Comment (timja) [inline pom.xml:473]: 🎉 thanks!

**Review by timja — APPROVED** [review]

**Review by oleg-nenashev — APPROVED** [review]
Looks good overall, thanks!

File: core/src/main/java/hudson/model/queue/FutureImpl.java
Comment (jglick) [reply core/src/main/java/hudson/model/queue/FutureImpl.java:67]: The class itself is `final` so this modifier is redundant.

**Review by jglick — COMMENTED** [review]
Did not read every line, but looked fine at a glance.

**Review by res0nance — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
